### ISSUE TITLE
Watched container vanishes if update takes longer than interval

### ISSUE THREAD

**Issue by ginkel** [containrrr/watchtower#417]
**Describe the bug**
I am using Watchtower to keep the Docker Containers in our system landscape up to date. Most servers are based on x86_64, some are ARM-based (Raspberry Pi 4).

During the update of some containers there is a high chance that the container vanishes after Watchtower has attempted to update it / updated it. In that case, the container is completely gone from `docker ps -a` output, i.e., it didn't just fail to start, but it completely vanished into thin air.

Home Assistant (`homeassistant/raspberrypi4-homeassistant:latest`) is such a Docker image / container.

The problem happened again last night and from the debug logs it seems that a contributing factor may be that the update of the Home Assistant container took longer than the Watchtower interval of 5 minutes (due to I/O on the Raspberry Pi being somewhat slow). I'm not 100% sure that this is the only cause, because I have only re-deployed Watchtower with `--debug` enabled yesterday evening.

This may be a duplicate of #414. 

**To Reproduce**
Tricky ;-)

Probably have Watchtower encounter an update of a container that takes longer than the configured interval to be updated, but that's solely based on my (still unvalidated) hypothesis.

**Expected behavior**
The container is updated by Watchtower and remains in existence afterwards.

**Screenshots**
n/a

**Environment**
 - Platform: Raspbian Linux 4.19.75-v7l+
 - Architecture: armv7l
 - Docker version: 19.03.5, build 633a0ea

**Logs from running watchtower with the `--debug` option**
```
time="2019-12-07T03:04:34Z" level=debug msg="Scheduled next run: 2019-12-07 03:09:13 +0000 UTC"
time="2019-12-07T03:09:13Z" level=debug msg="Checking containers for updated images"
time="2019-12-07T03:09:13Z" level=debug msg="Retrieving containers including stopped and exited"
time="2019-12-07T03:09:13Z" level=debug msg="Pulling containrrr/watchtower:armhf-latest for /watchtower"
time="2019-12-07T03:09:13Z" level=debug msg="No credentials for containrrr in /config.json"
time="2019-12-07T03:09:13Z" level=debug msg="Got auth value: "
time="2019-12-07T03:09:13Z" level=debug msg="Got image name: containrrr/watchtower:armhf-latest"
time="2019-12-07T03:09:13Z" level=debug msg="No authentication credentials found for containrrr/watchtower:armhf-latest"
time="2019-12-07T03:09:14Z" level=debug msg="No new images found for /watchtower"
time="2019-12-07T03:09:14Z" level=debug msg="Pulling homeassistant/raspberrypi4-homeassistant:latest for /homeassistant"
time="2019-12-07T03:09:14Z" level=debug msg="No credentials for homeassistant in /config.json"
time="2019-12-07T03:09:14Z" level=debug msg="Got auth value: "
time="2019-12-07T03:09:14Z" level=debug msg="Got image name: homeassistant/raspberrypi4-homeassistant:latest"
time="2019-12-07T03:09:14Z" level=debug msg="No authentication credentials found for homeassistant/raspberrypi4-homeassistant:latest"
time="2019-12-07T03:13:29Z" level=info msg="Found new homeassistant/raspberrypi4-homeassistant:latest image (sha256:47316baa78c5debb8d873a0b43e7865b8fee1db005db952ddee2b59caab3ccd6)"
[...]
time="2019-12-07T03:13:48Z" level=debug msg="No pre-update command supplied. Skipping"
time="2019-12-07T03:13:48Z" level=info msg="Executing pre-update command."
time="2019-12-07T03:13:48Z" level=info msg="Stopping /homeassistant (8e933e4f3ea9929b8a5651f1ed05f052606d69f34d91051462417be8e4f12723) with SIGTERM"
time="2019-12-07T03:14:13Z" level=debug msg="Skipped another update already running."
time="2019-12-07T03:14:13Z" level=debug msg="Scheduled next run: 2019-12-07 03:19:13 +0000 UTC"
time="2019-12-07T03:14:19Z" level=debug msg="Removing container 8e933e4f3ea9929b8a5651f1ed05f052606d69f34d91051462417be8e4f12723"
time="2019-12-07T03:14:19Z" level=debug msg="This is the watchtower container /watchtower"
time="2019-12-07T03:14:19Z" level=info msg="Creating /homeassistant"
time="2019-12-07T03:14:22Z" level=debug msg="Starting container /homeassistant (8c3a45a47cf52e09232137415f723deccab2935259bb2c3c548b6ae62fc7eed5)"
time="2019-12-07T03:14:25Z" level=debug msg="Scheduled next run: 2019-12-07 03:19:13 +0000 UTC"
time="2019-12-07T03:19:13Z" level=debug msg="Checking containers for updated images"
time="2019-12-07T03:19:13Z" level=debug msg="Retrieving containers including stopped and exited"
time="2019-12-07T03:19:13Z" level=debug msg="Pulling homeassistant/raspberrypi4-homeassistant:latest for /homeassistant"
time="2019-12-07T03:19:13Z" level=debug msg="No credentials for homeassistant in /config.json"
time="2019-12-07T03:19:13Z" level=debug msg="Got auth value: "
time="2019-12-07T03:19:13Z" level=debug msg="Got image name: homeassistant/raspberrypi4-homeassistant:latest"
time="2019-12-07T03:19:13Z" level=debug msg="No authentication credentials found for homeassistant/raspberrypi4-homeassistant:latest"
time="2019-12-07T03:19:14Z" level=debug msg="No new images found for /homeassistant"
[...]
```

**Additional context**

*Docker Journal*

```
Dec 07 03:13:48 lyra dockerd[1116]: time="2019-12-07T03:13:48.025597376Z" level=error msg="Error running exec acc268de56269750263fe707257cfe013360441516b90daca056f152908b3738 in container: Error: Exec command acc268de56269750263fe707257cfe013360441516b90daca056f152908b3738 is already running"
Dec 07 03:13:52 lyra dockerd[1116]: time="2019-12-07T03:13:52.794511850Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Dec 07 03:13:57 lyra dockerd[1116]: time="2019-12-07T03:13:57.605823464Z" level=warning msg="8e933e4f3ea9929b8a5651f1ed05f052606d69f34d91051462417be8e4f12723 cleanup: failed to unmount IPC: umount /var/lib/docker/containers/8e933e4f3ea9929b8a5651f1ed05f052606d69f34d91051462417be8e4f12723/mounts/shm, flags: 0x2: no such file or directory"
Dec 07 03:17:26 lyra dockerd[1116]: time="2019-12-07T03:17:26.003962382Z" level=info msg="Container 8c3a45a47cf52e09232137415f723deccab2935259bb2c3c548b6ae62fc7eed5 failed to exit within 180 seconds of signal 15 - using the force"
Dec 07 03:17:26 lyra dockerd[1116]: time="2019-12-07T03:17:26.568111858Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Dec 07 03:17:28 lyra dockerd[1116]: time="2019-12-07T03:17:28.405086341Z" level=warning msg="8c3a45a47cf52e09232137415f723deccab2935259bb2c3c548b6ae62fc7eed5 cleanup: failed to unmount IPC: umount /var/lib/docker/containers/8c3a45a47cf52e09232137415f723deccab2935259bb2c3c548b6ae62fc7eed5/mounts/shm, flags: 0x2: no such file or directory"
```

*Watchtower Version*

```
$ docker image ls
REPOSITORY                                              TAG                 IMAGE ID            CREATED             SIZE
[...]
containrrr/watchtower                                   armhf-latest        4229a81de7b7        2 months ago        12.5MB
```

*Watchtower Command-Line Options*

```
--debug
--include-stopped
--label-enable
--interval 300
--stop-timeout 30s
```

**Comment by simonaronsson** [comment]
Thank you for your report! 🙏🏼 I'll have a look as soon as I get some time on my hands :)

**Comment by simonaronsson** [comment]
Also, did you get any additional intel on this since you enabled `---debug`? 😄

**Comment by simonaronsson** [comment]
Ok, so I've had some time to look into this now, and I can't really tell what's going on. It should not be because of the interval being hit again during an update. If that's the case, we just reschedule the next check and continue with the one currently running. 

According to the watchtower logs at least, it seems like the homeassistant container is actually created successfully. If you run `docker ps -a`, is the homeassistant container present there perhaps (but exited)?

**Comment by Coriakin** [comment]
Also seeing this. Have not enabled debug yet though but containers do vanish without a trace and need to be manually recreated. It's especially annoying if it happens to my proxy containers which means watchtower cannot even notify me via gotify xD

I do not have any data to support my observation but I THINK I saw the same happening with "v2tec/watchtower", so I don't think it's unique to this watchtower. So maybe it's environmental based problem, or I suppose any code that is shared.

**Comment by ginkel** [comment]
I just got an interesting variation of this (on x86_64 as well as armhf): Updated image is detected, container is stopped / deleted, create fails b/c image unavailable:

```
time="2020-05-24T03:19:02Z" level=info msg="Found new tgbyte/smokeping:latest image (sha256:2434dbc14342eb03fbdb4ae228948600712b981e5bcea2578b4e7d0ed96cd690)"
[...]
time="2020-05-24T03:19:28Z" level=info msg="Creating /smokeping"
time="2020-05-24T03:19:28Z" level=error msg="Error response from daemon: No such image: tgbyte/smokeping:latest"
```

Needless to say that the image should have been available all the time...

**Comment by simonaronsson** [comment]
The first error seems to be related to #557 which was resolved (kind of) in version 1.0.1.

@ginkel - the error variation you're seeing should be solved by #528. If that's not the case and the image is indeed available, feel free to get back to us in a  separate issue.
